contracts/JBBuybackDelegate.sol

```diff
@@ -1,453 +1,538 @@
+            || _interfaceId == type(IJBBuybackDelegate).interfaceId
+            || super.supportsInterface(_interfaceId);
+    }
+
+    //*********************************************************************//
+    // ---------------------- external transactions ---------------------- //
+    //*********************************************************************//
+
+    /// @notice Delegate used to swap a provided amount to the beneficiary, using any leftover amount to mint.
+    /// @dev This delegate is called only if the quote for the swap is bigger than the quote when minting.
+    /// If the swap reverts (slippage, liquidity, etc), the delegate will then mint the same amount of token as if the delegate was not used.
+    /// @param _data The delegate data passed by the terminal.
     function didPay(JBDidPayData3_1_1 calldata _data) external payable override {
-        // Access control as minting is authorized to this delegate
+        // Make sure only a payment terminal belonging to the project can access this functionality.
         if (!DIRECTORY.isTerminalOf(_data.projectId, IJBPaymentTerminal(msg.sender))) {
             revert JuiceBuyback_Unauthorized();
         }
 
-        (uint256 _tokenCount, uint256 _swapMinAmountOut) = abi.decode(_data.dataSourceMetadata, (uint256, uint256));
-
-        // Try swapping
-        uint256 _amountReceived = _swap(_data, _swapMinAmountOut);
+        // Parse the metadata passed in from the data source.
+        (
+            bool _quoteExists,
+            bool _projectTokenIs0,
+            uint256 _minimumSwapAmountOut,
+            uint256 _weight
+        ) = abi.decode(_data.dataSourceMetadata, (bool, bool, uint256, uint256));
+
+        // Get a reference to the amount of tokens that was swapped for.
+        uint256 _exactSwapAmountOut = _swap(_data, _projectTokenIs0);
+
+        // Make sure the slippage is tolerable if passed in via an explicit quote.
+        if (_quoteExists && _exactSwapAmountOut < _minimumSwapAmountOut) revert JuiceBuyback_MaximumSlippage();
+
+        // Get a reference to any amount of tokens paid in remaining in this contract.
+        uint256 _terminalTokenInThisContract = _data.forwardedAmount.token == JBTokens.ETH
+            ? address(this).balance
+            : IERC20(_data.forwardedAmount.token).balanceOf(address(this));
+
+        // Use any leftover amount of tokens paid in remaining to mint.
+        // Keep a reference to the number of tokens being minted.
+        uint256 _partialMintTokenCount;
+        if (_terminalTokenInThisContract != 0) {
+            _partialMintTokenCount = mulDiv18(_terminalTokenInThisContract, _weight);
+
+            // If the token paid in wasn't ETH, give the terminal permission to pull them back into its balance.
+            if (_data.forwardedAmount.token != JBTokens.ETH) {
+                IERC20(_data.forwardedAmount.token).approve(msg.sender, _terminalTokenInThisContract);
```

Comment:
There is no chance of leftover token amounts, right? https://code4rena.com/reports/2023-05-juicebox#m-02-low-liquidity-in-uniswap-v3-pool-can-lead-to-eth-lockup-in-jbxbuybackdelegate-contract 

Also, what if this results in a large approval of JBX to a malicious terminal?

It does appear that all balances are settled atomically, but have we explored all the possibilities here?

Reply:
- addToBalance will pull _terminalTokenInThisContract (consuming the allowance)
- only a project owner can add a terminal for a given project (https://github.com/jbx-protocol/juice-contracts-v3/blob/5ae86dba2f27dee2dfe2235d20c3a070b7413ff6/contracts/JBDirectory.sol#L213) -> the check ln257 wouldn't pass if not